<a href="https://colab.research.google.com/github/bmcristin/Ingeniasrepo/blob/main/Notebook_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Programa Ingenias+ Data Science

## Trabajando con Pandas

1) Importa las librerias requeridas: `Pandas` y `NumPy`.  

Una buena practica a la hora de escribir codigo es de comentar siempre que estoy haciendo (usando `#`) en la linea anterior. Por mas obvio que parezca lo que voy a hacer, puede ser que tenga que compartir el documento con otras personas y no sepan que quise hacer en cada linea. O que yo mismo lea el codigo tiempo despues y no recuerde porque lo hice.

In [2]:
import numpy as np
import pandas as pd


In [3]:
# conecto my drive con collab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


2) Lee el archivo csv `life_expectancy_data` usando `pandas`. Guardalo en una variable llamada `expvida`.

In [4]:
# Leo los datos en un DataFrame
expvida = pd.read_csv("/content/drive/MyDrive/INGENIAS/notebook7/life_expectancy_data.csv")

3) ¿Que tipo de estructura de datos contiene la variable `expvida`?

In [5]:
# Obtengo el tipo de estructura de datos
print(type(expvida))

<class 'pandas.core.frame.DataFrame'>


4) ¿Cuantas filas y columnas tiene `expvida`?

In [6]:
# Uso .shape para obtener cantidad de filas y columnas
print(expvida.shape)

(2938, 22)


**Numero de filas**: 2938

**Numero de columnas**: 22

5) ¿Cual es el nombre de las columnas contenidas en `expvida`?

In [7]:
# Obtengo el nombre de las columnas
print(expvida.columns)

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


6) Inspecciona las primeras 10 filas de `expvida`

In [8]:
# Inspecciono las primeras filas
print(expvida.head(10))

       Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   
5  Afghanistan  2010  Developing              58.8            279.0   
6  Afghanistan  2009  Developing              58.6            281.0   
7  Afghanistan  2008  Developing              58.1            287.0   
8  Afghanistan  2007  Developing              57.5            295.0   
9  Afghanistan  2006  Developing              57.3            295.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582       

7) Y ahora las 10 ultimas.

In [9]:
# Inspecciono las ultimas filas
print(expvida.tail(10))

       Country  Year      Status  Life expectancy   Adult Mortality  \
2928  Zimbabwe  2009  Developing              50.0            587.0   
2929  Zimbabwe  2008  Developing              48.2            632.0   
2930  Zimbabwe  2007  Developing              46.6             67.0   
2931  Zimbabwe  2006  Developing              45.4              7.0   
2932  Zimbabwe  2005  Developing              44.6            717.0   
2933  Zimbabwe  2004  Developing              44.3            723.0   
2934  Zimbabwe  2003  Developing              44.5            715.0   
2935  Zimbabwe  2002  Developing              44.8             73.0   
2936  Zimbabwe  2001  Developing              45.3            686.0   
2937  Zimbabwe  2000  Developing              46.0            665.0   

      infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   \
2928             30     4.64                1.040021         73.0       853   
2929             30     3.56               20.843429        

8) ¿Que tipos de datos contiene cada una de las columnas de `expvida`?

In [10]:
# Tipo de datos en cada columna
print(expvida.dtypes)

Country                             object
Year                                 int64
Status                              object
Life expectancy                    float64
Adult Mortality                    float64
infant deaths                        int64
Alcohol                            float64
percentage expenditure             float64
Hepatitis B                        float64
Measles                              int64
 BMI                               float64
under-five deaths                    int64
Polio                              float64
Total expenditure                  float64
Diphtheria                         float64
 HIV/AIDS                          float64
GDP                                float64
Population                         float64
 thinness  1-19 years              float64
 thinness 5-9 years                float64
Income composition of resources    float64
Schooling                          float64
dtype: object


9) ¿Hay valores faltantes en alguna de las columnas de `expvida`?

In [11]:
# Obtengo la cantidad de datos faltantes en cada columna
print(expvida.isnull().sum())
# Si, hay valores faltantes donde la segunda columna es distanta de cero

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64


10) Selecciona el maximo de datos faltantes. Calcula que porcentaje de los datos representan los datos faltantes.

In [15]:
# Obtengo el maximo valor de datos faltantes y lo guardo en la variable masnan
masnan = expvida.isnull().sum()
columnaconmasnan = masnan.idxmax()
#calculo el porcentaje
totalnan = expvida.isnull().sum().sum()

porcentaje = (totalnan*100)/expvida.size
#imprimo
print(f"La columna con mas datos faltantes es \" {columnaconmasnan} \" con un total de NaN de : {masnan.max()}")
print(f"El porcentaje de datos nulos es {porcentaje} %")
print(columnaconmasnan)

La columna con mas datos faltantes es " Population " con un total de NaN de : 652
El porcentaje de datos nulos es 3.965282505105514 %
Population


#### Ahora vamos a realizar algunos calculos y manipulaciones sobre las columnas.

Una función útil para este fin es: `.rename()`. Y podemos utilizarla de la siguiente manera: Si `df` es mi DataFrame y quiero renombrar la columna `columna` como `columna1`, y la columna `columna2` como `columna3`, utilizo la siguiente sintaxis:
  
`df.rename(columns={"columna": "columna1", "columna2": "columna3"}, inplace=True)`

11) Renombra la columna 'Life expectancy ' como 'life_expectancy', ' BMI ' como 'bmi' y ''Measles ' como 'measles'. No olvides de usar `inplace` para modificar el DataFrame original. Luego chequea que el cambio se haya hecho correctamente.

In [ ]:
# Renombro columnas del DataFrame
expvida.rename(columns={'Life expectancy ':'life_expectancy', 'BMI ': 'bmi','Measles ': 'measles','Year':'year'}, inplace=True)

In [ ]:
# Chequeo que se hayan reemplazado correctamente
print(expvida.columns)

Index(['Country', 'year', 'Status', 'life_expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'measles', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')


12) ¿Cual es el promedio de expectativa de vida (Life expectancy) en general?

In [ ]:
# Obtengo el promedio de la columna
promedioLE = expvida['life_expectancy'].mean()
print(f"El promedio de vida es de {promedioLE} años")

El promedio de vida es de 69.22493169398908 años


13) ¿Cual es el promedio de expectativa de vida (Life expectancy) por país?

In [ ]:
# Obtengo el promedio por pais
promedioLEpais = expvida['life_expectancy'].groupby(expvida['Country']).mean()
print(promedioLEpais)

Country
Afghanistan                           58.19375
Albania                               75.15625
Algeria                               73.61875
Angola                                49.01875
Antigua and Barbuda                   75.05625
                                        ...   
Venezuela (Bolivarian Republic of)    73.38750
Viet Nam                              74.77500
Yemen                                 63.86250
Zambia                                53.90625
Zimbabwe                              50.48750
Name: life_expectancy, Length: 193, dtype: float64


14) ¿Cual es la expectativa de vida mas alta y mas baja? Usa la funcion de pandas y la funcion de NumPy y compara el resultado. ¿Que paises tienen esas expectativas de vida?

In [ ]:
# Maxima expectativa de vida usando numpy
maxLE = np.max(expvida.life_expectancy)
nombremaxLE = promedioLEpais.idxmax()
print(f'El pais con la mayor expectativa de vida es: {nombremaxLE} con una expectativa de vida de : {maxLE} años')

El pais con mayor expectativa de vida es: Japan con una expectativa de vida de : 89.0 años


In [ ]:
# Maxima expectativa de vida usando pandas
pdmaxLE = expvida.life_expectancy.max()
print(pdmaxLE)

89.0


In [ ]:
# Minima expectativa de vida usando numpy
minLE = np.min(expvida.life_expectancy)
nombreminLE = promedioLEpais.idxmin()
print(f'El pais con la minima expectativa de vida es: {nombreminLE} con una expectativa de vida de : {minLE} años')

El pais con minima expectativa de vida es: Sierra Leone con una expectativa de vida de : 36.3 años


In [ ]:
# Minima expectativa de vida usando pandas
pdminLE = expvida.life_expectancy.min()
print(pdminLE)

36.3


In [ ]:
# Muestro los resultados minimos


In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al maximo
colmaxLE = expvida[expvida.life_expectancy == maxLE]
print(colmaxLE)

In [ ]:
# Filtro las columnas para mostrar solo las filas que tengan valor
# igual al minimo
colminLE = expvida[expvida.life_expectancy == minLE]
print(colminLE)

15) ¿Cuantas categorias encontramos en la columna "Status"? _**Pista**_: Lee acerca de la función [unique()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.unique.html?highlight=unique#pandas.Series.unique)

In [ ]:
# Obtengo los valores unicos de la columna status
difstatus = expvida['Status'].unique()
print(difstatus)

['Developing' 'Developed']


In [ ]:
# Cuento cuantos valores hay de cada categoria
developed = np.count_nonzero(expvida['Status'] == 'Developed')
developing = np.count_nonzero(expvida['Status'] == 'Developing')
print(f'La cantidad de paises desarrollados es: {developed}')
print(f'La cantidad de paises en desarrollo es: {developing}')

La cantidad de paises desarrollados es: 512
La cantidad de paises en desarrollo es: 2426


18) ¿Cuantos paises presentaron algun caso de Sarampion (**Measles**) en el 2015? **_Pista_**: Lee acerca de la función [.count()](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html?highlight=count#pandas.DataFrame.count)

In [ ]:
# Cuento cuantas filas tienen valores distintos de cero
paisesSarampion = expvida[(expvida.measles != 0) & (expvida.year == 2015)].count()
print(paisesSarampion)

Country                            121
year                               121
Status                             121
life_expectancy                    121
Adult Mortality                    121
infant deaths                      121
Alcohol                              4
percentage expenditure             121
Hepatitis B                        114
measles                            121
 BMI                               119
under-five deaths                  121
Polio                              121
Total expenditure                    1
Diphtheria                         121
 HIV/AIDS                          121
GDP                                100
Population                          95
 thinness  1-19 years              119
 thinness 5-9 years                119
Income composition of resources    113
Schooling                          113
dtype: int64
